Feature email video article published

gnott opened the pull request:
In reference to issue https://github.com/elifesciences/elife-bot/issues/561

Adds a new activity named `EmailVideoArticlePublished` that will email recipients once only (no duplicates allowed) if an article that contains a video is published. The recipients are specified in the bot settings, and they are internal recipients, not the authors of the article.

A little refactoring in the downloading of templates in the templates provider.

I also added some methods to the activity object in `activity.py` for emitting some messages which many of the activities do. These could be used later to refactor the message sending in other activities. I think they are general purpose enough to use the common methods.

There are additional bot settings to be added for this before it is deployed. I also need to add the new email templates before deployment.

giorgiosironi commented on activity/activity.py:
```diff
@@ -198,6 +198,28 @@ def clean_tmp_dir(self):
+        "emit the start message to the queue"
+        return self.emit_activity_message(
+            article_id, version, run, "start",
+            "Starting ingest digest to endpoint for " + str(article_id))
```
these messages look too specific (digests-related)

gnott replied:
Oops, I will fix the basics to try and make it suitable for general use.

giorgiosironi commented on activity/activity_EmailVideoArticlePublished.py:
```diff
@@ -0,0 +1,315 @@
+        self.description = "Send email when an article containing a video is published."
+
+        # Data provider
+        self.database = dblib.SimpleDB(settings)
```
adding dependencies on SimpleDB at this stage is a strategic problem, any other source of data and/or storage of state that can be used?

giorgiosironi replied:
> It is not a complicated situation, but it requires effort and care.

Definitely, let's not add even more variations, as you say.

gnott replied:
I'm having a rethink on the duplicates. We might be able to base it on sending emails

- only on first VoR published
- never after a silent correction

Then the duplicate check doesn't need to check the sending history.

giorgiosironi commented on activity/activity_EmailVideoArticlePublished.py:
```diff
@@ -0,0 +1,315 @@
+                                  " digest to endpoint. Details: %s" % str(exception))
+        return success, run, article_id, version, status, expanded_folder
+
+    def download_xml(self, expanded_bucket, expanded_folder, version, to_dir):
```
this function seems to achieve a very common operation; should be extracted into a provider (`xml_provider` using `storage_provider` underneath?)

giorgiosironi replied:
This is even more interesting for the email-related code, what can be pushed into a `email_provider` rather than being rewritten in each activity? There will be bits dependent on settings or on particular session data, but the general machinery of rendering and sending an email should be reusable

gnott replied:
There's an opportunity for refactoring here. This particular activity is taking the input from the expanded folder of the article ingest workflow.

 The publication emails use an S3 outbox system for input, and is separate from the article ingest workflow, so it wouldn't apply there. 

The best place for a shared function might be in the `execution_context.py` or maybe `article_structure.py`, or a new provider.

There are a few duplicates of this action, some older and some slightly different. The activities `ApplyVersionNumber` and `ModifyArticleSubjects` download the XML from the expanded folder. There are also similar functions for finding the article XML in the expanded folder in `article_structure/get_article_xml_key()` (https://github.com/elifesciences/elife-bot/blob/develop/provider/article_structure.py#L189) and `article/get_xml_file_name()` (https://github.com/elifesciences/elife-bot/blob/develop/provider/article.py#L734), also used by `lax_provider.get_xml_file_name()` (https://github.com/elifesciences/elife-bot/blob/develop/provider/lax_provider.py#L217).

No one has attempted to find the commonality between all of these and replace them in all the activities yet. It has evolved over time with different developers.

Trying to refactor this might be better attempted on a separate branch and a different PR for the functions listed above. This new `EmailVideoArticlePublished` activity could be blocked by that refactoring, and then once merged it can be used here.

What do you think about a separate refactor for finding and downloading article XML from the expanded bucket?

giorgiosironi replied:
> about a separate refactor for finding and downloading article XML from the expanded bucket

Definitely good to have an issue for that, it affects lots of key activities; whereas emails do not have the same impact on publication (I think).

giorgiosironi commented on activity/activity_EmailVideoArticlePublished.py:
```diff
@@ -0,0 +1,315 @@
+
+    def is_duplicate_email(self, doi_id, email_type, recipient_email):
+        """
+        Use the SimpleDB provider to count the number of emails
```
where do these duplicates come from? Is the activity executed multiple times for a single publishing workflow?

gnott replied:
The SimpleDB records form the history of emails, and duplicates are found in that history. It checks duplicates by a combination of `email_template_name` + `recipient_email`.

giorgiosironi commented on activity/activity_EmailVideoArticlePublished.py:
```diff
@@ -0,0 +1,315 @@
+            format=email_format)
+
+        # Add the email to the email queue
+        self.database.elife_add_email_to_email_queue(
```
no idea what happens once the emails get into a queue in the database. Do they get sent immediately by some other process? By a cron? Are they scheduled for later?

gnott replied:
All queued emails (the queue is stored in SimpleDB) are sent in the `SendQueuedEmail` workflow, the cron starts every 5 minutes. They are immediate in this case (but the queue does support scheduling emails to be sent at a future date).

This new `EmailVideoArticlePublished` activity doesn't need to rely on the email queue, if it looks like a non-SimpleDB method of history and persistence is a better option here. We could track history of sending on a per-article basis by keeping track in S3 folders, and we could also send emails immediately using the SMTP method we know have when it was created for sending digest content.

giorgiosironi replied:
There are already a couple of queues (Postfix/SES) in place if you use SMTP; so if the queuing can be dropped without having to implement persistence with S3 and sending emails directly, I'd go for that. Assuming that removes code or executes less existing code, which is why it would be helpful as fewer things can break/need to be maintained.

gnott replied:
If the logic to send it based on VoR version and `run_type`, then there's no need to add it to the queue to check duplicates. We can just send directly by the SMTP method.

gnott commented:
It's been a while for this PR. I addressed earlier review comments.

Recent changes were to switch over to sending email direct by SMTP (instead of the older email queue), and to change how it decides when to send an email (no longer relying on checking for duplicate emails).

Also refactored how the XML is downloaded from the expanded folder, using `article_processing` provider now.

I expanded the `email_provider.py` to load email headers and body from templates with exception handling. It makes this activity code simpler, and is useful in further email refactoring I'm planning soon.

Including @GiancarloFusiello for review too if you are curious.

All the boxes I self-defined on issue https://github.com/elifesciences/elife-bot/issues/561 are completed here.

Thanks!

gnott commented:
Thanks for taking a look again, I guess if tests are passing it is time to merge and watch it go!
